Add Kubernetes-in-Docker based automated tests

diverdane opened the pull request:
This change adds automated testing that uses Kubernetes-in-Docker (KinD)
to spin up Kubernetes clusters using various Kubernetes versions.

Currently, the Kubernetes versions being used for testing are:
- Kubernetes v1.18.2
- Kubernetes v1.16.9
- Kubernetes v1.14.10
- Kubernetes v1.12.10

The tests that are run for each version of Kubernetes include:
- KinD create of a Kubernetes cluster.
- Helm install of Conjur/Nginx + backend Postgres database.
- Helm test of the Conjur deployment that confirms that the
  Conjur status page is active.
- Helm delete of Conjur release.
- KinD delete of the Kubernetes cluster.

This change also includes some Helm v2 -> Helm v3 syntax changes, since Helm v3
is our officially supported version. Also, github action test is modified to use Helm v3.2.

Also included is the addition of a new chart values setting to indicate whether the
test pods should be automatically deleted when the helm test passes. From values.yaml:
```
test:
  deleteOnSuccess: true  # If true, test pods are deleted upon successful
                         # test completion. Otherwise, test pods are not
                         # automatically deleted upon test completion,
                         # regardless of test success or failure.
```
This defaults to `true`, but should be set to `false` e.g. when `helm test` logging is enabled,
since `helm test` needs to display the test pod log before the pod gets deleted, even for
test success.

Fixes Issue #52

izgeri commented:
I think this also resolves #32 

what's the benefit of doing this in jenkins rather than github actions? does the pipeline require privileges of any kind?

diverdane commented:
> what's the benefit of doing this in jenkins rather than github actions? does the pipeline require privileges of any kind?

Honestly, I'm not sure what the pros/cons of running this via jenkins rather than github actions.
I started with a Jenkins approach because I was more familiar with that approach, and had some
examples to start with.
For privileged mode, that's a good question... I run the test client container in not-privileged mode, and I think KinD runs non-privileged containers as well. But I am fairly certain that we'll need to run in privileged mode if we ever want to test Conjur on an IPv6 cluster.

izgeri commented:
re: privilege - I'm not referring to the way the container runs as much as whether it needs access to privileged systems (like our internal infra, our logins to publish artifacts, etc)

if it doesn't need access to any of our privileged systems, it would be nice to run the tests with gh actions since community users can (in theory) see the PR feedback - they can't in jenkins. if that's not straightforward to do, though, no need to add it onto this PR, which adds a ton of value just by adding automated tests back into this repo. we may want to consider it in the future, though.

sgnn7 commented on Jenkinsfile:
```diff
@@ -0,0 +1,106 @@
+    parallel {
+
+      # For available Docker tags for Kubernetes node images see:
+      #     https://github.com/kubernetes-sigs/kind/releases
```
I think groovy script comments are c-style comments: https://docs.groovy-lang.org/docs/next/html/documentation/#_comments

sgnn7 commented on Jenkinsfile:
```diff
@@ -0,0 +1,106 @@
+import groovy.transform.Field
+
+@Field
+def TAG = ""
```
Might be due to lack of my pipeline understanding but why can't this be a toplevel `environment` directive instead?

sgnn7 commented on Jenkinsfile:
```diff
@@ -0,0 +1,106 @@
+          KUBE_NODE_TAG   = "v1.18.2@sha256:7b27a6d0f2517ff88ba444025beae41491b016bc6af573ba467b70c5e8e0d85f"
+          KUBECTL_VERSION = "1.18.2"
+          CLUSTER_ID      = "1.18.2"
+          GIT_REF = sh(returnStdout: true, script: 'echo $(< VERSION)-$(git rev-parse --short HEAD) | tr -d "\n"')
```
Can this variable be evaluated beforehand for all subsequent usages of it?

sgnn7 commented on Jenkinsfile:
```diff
@@ -0,0 +1,106 @@
+            KUBE_NODE_TAG   = "${env.KUBE_NODE_TAG}"
+            KUBECTL_VERSION = "${env.KUBECTL_VERSION}"
+            CLUSTER_ID      = "${env.CLUSTER_ID}"
+            GIT_REF         = "${env.GIT_REF}"
```
I'm leaning a little towards not explicitly defining `environment` for these since you're just directly using these variables right afterwards in a single spot each time.

sgnn7 commented on ci/jenkins_build.sh:
```diff
@@ -0,0 +1,64 @@
+echo "==================================="
+
+KIND_VERSION=0.7.0
+export CLUSTER_NAME="${CLUSTER_ID}-${GIT_REF}"
```
It would be good to see the `export`ed variables grouped in a single spot instead of spread around the file

sgnn7 commented on ci/jenkins_build.sh:
```diff
@@ -0,0 +1,64 @@
+             -f ${CURRENT_DIR}/Dockerfile \
+             --build-arg KIND_VERSION="${KIND_VERSION}" \
+             --build-arg KUBECTL_VERSION="${KUBECTL_VERSION}" \
+             "${TOPLEVEL_DIR}"
```
Since this is just a test but a long and parallel one, I wonder if we can try using `DOCKER_BUILDKIT=1` for faster runs.

diverdane replied:
Ooh, yeah, this is cool. I like the new build output format too.

sgnn7 commented on ci/test_runner.sh:
```diff
@@ -0,0 +1,84 @@
+nodes:
+- role: control-plane
+  image: "kindest/node:$KUBE_NODE_TAG"
+EOT
```
Add a newline at the end of each `EOT` and/or at the beginning of `if`s - it's hard to read this as a single block

sgnn7 commented on ci/test_runner.sh:
```diff
@@ -0,0 +1,84 @@
+#      <Master Node Container's IP address>:6443
+echo "Fixing up Kubernetes config file..."
+MASTER_IP=$(docker inspect --format='{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' ${CLUSTER_NAME}-control-plane)
+echo "Master Node IP Address: ${MASTER_IP}"
```
nit: Add a newline above this line

sgnn7 commented on ci/test_runner.sh:
```diff
@@ -0,0 +1,84 @@
+echo "Fixing up Kubernetes config file..."
+MASTER_IP=$(docker inspect --format='{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' ${CLUSTER_NAME}-control-plane)
+echo "Master Node IP Address: ${MASTER_IP}"
+sed -i "s/^    server:.*/    server: https:\/\/$MASTER_IP:6443/" /root/.kube/config
```
not a blocker but this seems pretty fragile

diverdane replied:
Ooh, yes, very fragile. :o
I changed this to use the `kubectl config set clusters.<cluster-name>.server <server-address` command.

sgnn7 commented on test.sh:
```diff
@@ -1,24 +1,43 @@
 
-HELM_TEST_ARGS="${@:---cleanup}"  # cleanup test pod by default
+# Command line arguments for this script are passed to `helm test`.
+HELM_TEST_ARGS="$@"
```
Hmm... Do we not want cleanup by default?

sgnn7 commented on test.sh:
```diff
@@ -1,24 +1,43 @@
+             --timeout 900s \
+             --set "dataKey=$dataKey" \
+             $HELM_INSTALL_ARGS \
+             $RELEASE_NAME \
```
This should be quoted

sgnn7 commented on test.sh:
```diff
@@ -1,24 +1,43 @@
+  echo "=============================="
+  echo "Deleting release $RELEASE_NAME"
+  echo "=============================="
+  helm del $RELEASE_NAME
```
This should be quoted

sgnn7 commented on ci/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+FROM alpine:3.8 as conjur-test
```
is there a reason not to use `alpine:latest` or at least `apline:3.11`?

diverdane replied:
I moved the testing from Jenkins to github actions, so Dockerfile for test client pod was deleted.

sgnn7 commented on ci/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+# Install Kubernetes in Docker (kind)
+RUN curl -Lo ./kind https://github.com/kubernetes-sigs/kind/releases/download/v${KIND_VERSION}/kind-linux-amd64 && \
+    chmod +x ./kind && \
+    mv ./kind /usr/local/bin/kind
```
I just noticed this, but can't you just directly download the file into `/usr/local/bin`? seems like download + move is doing double the work but I can't tell why.

diverdane replied:
This file has been deleted, but yeah, that would have made sense!

sgnn7 commented on ci/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+    chmod 700 get_helm.sh && \
+    ./get_helm.sh
+
+#COPY conjur-oss /root/conjur-oss/
```
Needs removal?

diverdane replied:
Yes. But now the entire file has been deleted.

sgnn7 commented on ci/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+
+ENV PATH="${PATH}:/root"
+
+WORKDIR /root
```
Is `/root` not the default workdir? If it is, this needs deleting and if it's not it needs to move to the top of the file.

diverdane replied:
Good point... but this Dockerfile is no longer needed.

sgnn7 commented on ci/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+
+COPY test.sh ci/test_runner.sh /root/
+
+ENV PATH="${PATH}:/root"
```
Are just test runners needed in root? Seems a bit odd to need this.

diverdane replied:
Yeah, probably should have moved these to /usr/local/bin or somewhere in the path. But Dockerfile isn't needed now.

diverdane commented on ci/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+#COPY conjur-oss /root/conjur-oss/
+COPY conjur-oss /root/conjur-oss/
+
+COPY test.sh ci/test_runner.sh /root/
```
Roger that. But Dockerfile is no longer needed.

diverdane commented:
@sgnn7 : Thank you for the reviews! 
I've switched the testing from Jenkins to Github actions. This GREATLY simplifies things.
For the Github actions, I used a `matrix` strategy to test against the various Kubernetes
versions. This causes the tests for the different versions to run in parallel.

diverdane commented:
@sgnn7 I've added some last-minute `Helm V2 vs V3` changes:
* Added a function to detect Helm version, and use function that to select between Helm V2 and V3 syntax at various points in the scripts.
* Added a Helm V2 github action test (in addition to the Helm v3 tests previously added). The Helm v2 test has to be run sequentially for some reason, I think there are Github action resource limitations when you run more than 4 KinD clusters in parallel???

diverdane commented on .github/workflows/ci.yml:
```diff
@@ -5,8 +5,49 @@ on:
+
+  install-test-helm-v2:
+    name: Install/test Conjur with Helm V2 on KinD Cluster
+    needs: [ linter,install-test-helm-v3 ]
```
Yes, good suggestion. Fixed.

diverdane commented on is_helm_v2.sh:
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
```
Oops, adding the shebang is so automatic. :) Fixed.

diverdane commented on is_helm_v2.sh:
```diff
@@ -0,0 +1,16 @@
+  # option. For Helm version 3 or newer, the release name is a positional
+  # argument. (The versions can be distinguished by whether or not
+  # 'helm version --server' is supported.)
+  helm version --server > /dev/null 2>&1 && returncode=$? || returncode=$?
```
Fixed.

diverdane commented on is_helm_v2.sh:
```diff
@@ -0,0 +1,16 @@
+    echo true
+  else
+    echo false
+  fi
```
Yes, much cleaner!

diverdane commented on run.sh:
```diff
@@ -1,10 +1,18 @@
   HELM_ARGS="--set dataKey=$data_key"
 fi
 
+if [[ "$(is_helm_v2)" == true ]]; then
```
Yes, great simplification!
